draft/XXXX-robust-api/README.md

```diff
@@ -0,0 +1,146 @@
+
+
+### Finalize the upload
+
+`POST /api/workspaces/{workspace}/upload/finalize`
+
+This is the last step that the client is responsible for. This instructs the server to finish the upload. The payload for the request must contain the following:
+
+```
+{
+  "uploadPartTags": [
+    <ETag returned from the first signed url>,
+    <ETag returned from the second signed url>,
+    ...
+  ]
+}
+```
+
+**NOTE:** Similar to the `uploadPartUrls` array mentioned in the Upload creation step, the ordering of the `uploadPartTags` array is significant.
+
+At this point, the sever downloads the reconstructed file from S3, and performs data processing, table/graph creation, etc. with the appropriate processor for that data type. This process doesn't technically need to be done asynchronously, as if the data is small enough it's feasible that this could be completed within the time frame of a single request. However, in the interest of scalability and robustness, this HTTP request should initiate an asynchronous task to process the data.
+
+## Additional Changes
+
+There are changes in addition to the endpoints mentioned above, which may not be crucial to the client, but should be detailed.
+
+### Upload Collection
+
+Each workspace should contain its own upload collection, which will be used to store each upload document shown above. This will need to be accompanied by a new `Upload` ORM, which will facilitate operations on uploads.
+
+### Additional Upload Endpoints
+
+`GET /api/workspaces/{workspace}/upload/{id}`
+
+This endpoint returns the Upload document with the corresponding ID.
+
+## Open Questions
+
+### Changes to data processing procedure
+
+The notion of having various file type "uploaders" will be removed. Instead, these will become data processors, which will take raw uploaded data (from S3), and run validation, table/graph creation, data insertion, etc. When downloading the finalized upload from S3, this processing should be done in a streaming manner, in order to limit the risk of running out of memory in the server. This will involve updating each data processor to handle streamed data, instead of the entire file. This could be done in the following manner:
+
+1. Read in data until a fully formed object can be parsed (a table row, for example).
+2. Run validation on this object
+3. If validation is passed, insert this object into the target location. If validation fails, save the validation error, and either continue processing more data, or halt data processing.
+4. Repeat steps 1-3 until all data has been processed
+5. If there are errors resulting from this processing, delete the partially formed tables/graphs, and return the validation errors to the user.
+
+These errors would likely be stored in the upload model (yet to be outlined), however, it may make sense to introduce another model that would track this status more specifically. Feedback is welcome here.
```

Comment:
We need to design this to complete within the Heroku limitations of 30 seconds per part, too. Although 1 row probably wouldn't cause us any problems

Reply:
Or, as mentioned above, we could use async for this

Reply:
Yeah imo this should be done asynchronously using something like celery, even though for small enough data we could probably get away with it.

Reply:
That everything isn't "small enough data" is precisely the problem we're trying to solve. Go ahead and make the assumption that we need async task support for this.